File: 2067-maximum-number-of-points-with-cost/solution.py
class Solution:
    def maxPoints(self, points: List[List[int]]) -> int:
        rows, cols = len(points), len(points[0])
        previous_row = points[0]

        for row in range(1, rows):
            left_max    = [0 for _ in range(cols)]
            right_max   = [0 for _ in range(cols)]
            current_row = [0 for _ in range(cols)]
            left_max[0] = previous_row[0]
            for col in range(1, cols):
                left_max[col] = max(previous_row[col], left_max[col-1]-1)
            
            right_max[-1] = previous_row[-1]
            for col in range(cols - 2, -1, -1):
                right_max[col] = max(right_max[col+1] - 1, previous_row[col])
            
            for col in range(cols):
                current_row[col] = max(left_max[col], right_max[col]) + points[row][col]
            
            previous_row = current_row
        
        return max(previous_row)

File: 2067-maximum-number-of-points-with-cost/solution.cpp
class Solution {
public:
    long long maxPoints(vector<vector<int>>& points) {
        vector<long long> row(points[0].begin(), points[0].end());
        int num_rows = points.size(), num_cols = points[0].size();

        vector<long long> left_max(num_cols, 0);
        vector<long long> right_max(num_cols, 0);
        vector<long long> cur_row(num_cols, 0);

        for (int row_idx = 1; row_idx < points.size(); row_idx++) {
            left_max.assign(left_max.size(), 0);
            right_max.assign(right_max.size(), 0);


            left_max[0] = row[0];            
            for (int col_idx = 1; col_idx < num_cols; col_idx++) {
                left_max[col_idx] = 
                    max(left_max[col_idx - 1] - 1, row[col_idx]);
            }

            right_max.back() = row.back();
            for (int col_idx = num_cols - 2; col_idx >= 0; col_idx--) {
                right_max[col_idx] = 
                    max(right_max[col_idx + 1] - 1, row[col_idx]);
            }

            for (int col_idx = 0; col_idx < num_cols; col_idx++) {
                cur_row[col_idx] = points[row_idx][col_idx];
                cur_row[col_idx] += max(left_max[col_idx], right_max[col_idx]);
            }
            row = cur_row;
        }

        return *max_element(row.begin(), row.end());
    }
};
